gtsam/hybrid/DCSAM.h

```diff
@@ -0,0 +1,202 @@
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    DCSAM.h
+ * @brief   Discrete-Continuous Smoothing and Mapping for Factored Models
+ * @author  Kevin Doherty
+ * @author  Varun Agrawal
+ * @date    March 2025
+ */
+
+#pragma once
+
+#include <gtsam/discrete/DiscreteFactor.h>
+#include <gtsam/discrete/DiscreteFactorGraph.h>
+#include <gtsam/discrete/DiscreteKey.h>
+#include <gtsam/discrete/DiscreteMarginals.h>
+#include <gtsam/hybrid/HybridNonlinearFactorGraph.h>
+#include <gtsam/inference/FactorGraph.h>
+#include <gtsam/inference/Symbol.h>
+#include <gtsam/nonlinear/ISAM2.h>
+#include <gtsam/nonlinear/Marginals.h>
+#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+
+#include <map>
+#include <set>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+namespace gtsam {
+
+class GTSAM_EXPORT DCSAM {
+ private:
+  // Global factor graph and iSAM2 instance
+  NonlinearFactorGraph nfg_;
+  DiscreteFactorGraph dfg_;
+  ISAM2 isam_;
+  Values currContinuous_;
+  DiscreteValues currDiscrete_;
+
+ public:
+  DCSAM();
+
+  explicit DCSAM(const ISAM2Params &isam_params);
+
+  /**
+   * For this solver, runs an iteration of alternating minimization between
+   * discrete and continuous variables, adding any user-supplied factors (with
+   * initial guess) first.
+   *
+   * 1. Adds new discrete factors (if any) as supplied by a user to the
+   * discrete factor graph, then adds any discrete-continuous factors to the
+   * discrete factor graph, appropriately initializing their continuous
+   * variables to those from the last solve and any supplied by the initial
+   * guess.
+   *
+   * 2. Update the solution for the discrete variables.
+   *
+   * 3. For all new discrete-continuous factors to be passed to the continuous
+   * solver, update/set the latest discrete variables (prior to adding).
+   *
+   * 4. In one step: add new factors, new values, and earmarked old factor keys
+   * to iSAM. Specifically, loop over DC factors already in iSAM, updating their
+   * discrete information, then call isam_.update() with the (initialized) new
+   * DC factors, any new continuous factors, and the initial guess to be
+   * supplied.
+   *
+   * 5. Calculate the latest continuous variables from iSAM.
+   *
+   * 6. Update the discrete factors in the discrete factor graph dfg_ with the
+   * latest information from the continuous solve.
+   *
+   * @param graph - a HybridNonlinearFactorGraph containing all factors to add.
+   * @param initialGuessContinuous - an initial guess for any new
+   * continuous keys that appear in the updated factors
+   * (or if one wants to force override previously obtained continuous values).
+   * @param initialGuessDiscrete - Initial guess for discrete variables.
+   */
+  void update(const HybridNonlinearFactorGraph &graph,
+              const Values &initialGuessContinuous = Values(),
+              const DiscreteValues &initialGuessDiscrete = DiscreteValues());
```

Comment:
I wonder if the new hybrid machinery could be used to simplify this? Is there some kind of `HybridValues` (or equivalent?)

Reply:
Yes there is!!